src/transaction.ts

```diff
@@ -0,0 +1,205 @@
+    throw new Error('satsPerByte minimum value is 0.1');
+  }
+
+  if (args.recipients.length === 0) {
+    throw new Error(
+      'need a least one recipient output to build the transaction'
+    );
+  }
+
+  if (args.unspents.length === 0) {
+    throw new Error('need at least one unspent to fund the transaction');
+  }
+
+  return args;
+}
+
+/**
+ * buildTx selects utxos among unspents to fill outputs' requirements,
+ * then it adds the associated inputs and outputs to the tx.
+ * finally it returns the new pset base64 encoded.
+ * @param psetBase64
+ * @param unspents
+ * @param outputs
+ * @param changeScriptByAsset
+ * @param addFee if true, estimate fee and add feeOutput to transaction (default = false)
+ * @param satsPerByte used for fee estimation (default = 0.1)
+ * @param network used for fee output (default = regtest)
+ */
+export function buildTx(args: BuildTxArgs): string {
+  // validate and deconstruct args object
+  const {
+    changeAddressByAsset,
+    coinSelector,
+    psetBase64,
+    recipients,
+    unspents,
+    addFee,
+    network,
+    satsPerByte,
+  } = validateAndProcess(args);
+
+  const { selectedUtxos, changeOutputs } = coinSelector(
+    unspents,
+    recipients,
+    changeAddressByAsset
+  );
+
+  const inputs = selectedUtxos;
+
+  // if not fee, just add selected unspents as inputs and specified outputs + change outputs to pset
+  if (!addFee) {
+    const outs = recipients.concat(changeOutputs);
+    return addToTx(psetBase64, inputs, outs, network!);
+  }
+
+  const pset = decodePset(psetBase64);
+  const nbInputs = pset.data.inputs.length + inputs.length;
+  let nbOutputs =
+    pset.data.outputs.length + recipients.length + changeOutputs.length;
+
+  // otherwise, handle the fee output
+  const fee = createFeeOutput(nbInputs, nbOutputs, satsPerByte!, network!);
+
+  const changeIndexLBTC: number = changeOutputs.findIndex(
+    out => out.asset === network!.assetHash
+  );
+
+  let diff =
+    changeIndexLBTC === -1
+      ? 0 - fee.value
+      : changeOutputs[changeIndexLBTC].value - fee.value;
+
+  if (diff > 0) {
+    changeOutputs[changeIndexLBTC].value = diff;
+    const outs = recipients.concat(changeOutputs).concat(fee);
+    return addToTx(psetBase64, inputs, outs, network!);
+  }
+  // remove the change outputs (if it exists)
+  if (changeIndexLBTC > 0) {
+    changeOutputs.splice(changeIndexLBTC, 1);
+    nbOutputs -= 1;
+  }
+
+  if (diff === 0) {
+    const outs = recipients.concat(changeOutputs).concat(fee);
+    return addToTx(psetBase64, inputs, outs, network!);
+  }
+
+  const availableUnspents: UtxoInterface[] = [];
+  for (const utxo of unspents) {
+    if (!selectedUtxos.includes(utxo)) availableUnspents.push(utxo);
+  }
+
+  // re-estimate the fees with one additional output
+  const feeBis = createFeeOutput(
+    nbInputs + 1,
+    nbOutputs,
+    satsPerByte!,
+    network!
+  );
+
+  // reassign diff to new value
+  diff = fee.value - feeBis.value + diff;
```

Comment:
couldn't this just be `diff += feeBis.value`?

Reply:
it is, thanks